x/ibc/07-tendermint/types/proposal_handle.go

```diff
@@ -0,0 +1,101 @@
+package types
+
+import (
+	"time"
+
+	tmtypes "github.com/tendermint/tendermint/types"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	clientexported "github.com/cosmos/cosmos-sdk/x/ibc/02-client/exported"
+	clienttypes "github.com/cosmos/cosmos-sdk/x/ibc/02-client/types"
+)
+
+// CheckProposedHeaderAndUpdateState will try to update the client with the new header if and
+// only if the proposal passes and one of the following two conditions is satisfied:
+// 		1) AllowGovernanceOverrideAfterExpiry=true and Expire(ctx.BlockTime) = true
+// 		2) AllowGovernanceOverrideAfterMinbehaviour and IsFrozen() = true
+// In case 2) before trying to update the client, the client will be unfrozen by resetting
+// the FrozenHeight to the zero Height. Note, that even if the update happens, it may not be
+// successful. The header may fail validation checks and an error will be returned in that
+// case.
+func (cs ClientState) CheckProposedHeaderAndUpdateState(
+	ctx sdk.Context, cdc codec.BinaryMarshaler, clientStore sdk.KVStore,
+	header clientexported.Header,
+) (clientexported.ClientState, clientexported.ConsensusState, error) {
+	tmHeader, ok := header.(*Header)
+	if !ok {
+		return nil, nil, sdkerrors.Wrapf(
+			clienttypes.ErrInvalidHeader, "expected type %T, got %T", &Header{}, header,
+		)
+	}
+
+	// get consensus state corresponding to client state to check if the client is expired
+	consensusState, err := GetConsensusState(clientStore, cdc, cs.GetLatestHeight())
+	if err != nil {
+		return nil, nil, sdkerrors.Wrapf(
+			err, "could not get consensus state from clientstore at height: %d", cs.GetLatestHeight(),
+		)
+	}
+
+	// unfreeze client if the client is frozen and this is allowed. If the client is expired
+	// and allowed to be updated after expiry then we use do light validation on the header.
+	// Otherwise we process the header as normal to update the client. If none of these
+	// conditions are met then an error is returned.
+	if cs.IsFrozen() && cs.AllowGovernanceOverrideAfterMisbehaviour {
```

Comment:
There needs to be some additional logic in this case to prevent refreezing. If the override happens and the misbehaviour is still within the unbonding period, then it can simply be resubmitted and freeze the chain again.

You probably want to create an additional field in client state, `BaseTrustedHeight` or something and prevent any misbehaviour being submitted before it in `misbehavior_handle.go`. Initialize this to the first submitted header's height and update on any override since you are essentially using the new override header as the "base of trust" for your light client moving forward

Reply:
Hmm, we could do this; the user can just wait until the misbehaviour has expired to update the client with a governance proposal, although I suppose it's annoying. I think your proposed solution is fine though.